### ISSUE TITLE
Fix location of hashValue implementation?

### ISSUE THREAD

**Issue by allevato** [apple/swift-protobuf#233]
Is there a reason that `hashValue` is implemented as an extension on `Message`, rather than an extension on `_MessageImplementationBase` where the `Hashable` conformance is defined? Should we move it?

**Comment by dpassage** [comment]
Having `hashValue` be an extension on `Message` makes it appear in `Message`'s documentation. Since it's done as an extension intentionally so users can override it, that's probably the best place to document it.

Maybe `Hashable` should be on `Message` instead?

**Comment by tbkka** [comment]
`Hashable` cannot be on `Message` because it prevents `Message` from being used as a variable type.

**Comment by allevato** [comment]
@dpassage Good point about the documentation aspect—that provides some value. That being said, users also have the ability to provide their own implementation of `==` if they wish, and we can't hoist that into the top protocol because of the associated type constraint.

It's not a major bug—just one of the "huh?"s that I came across when I was re-reviewing the code recently. Until Swift supports better treatment of existentials, we may have to accept this as it is right now (of course, at that time we would be able to rid ourselves of `_MessageImplementationBase` entirely).
